### ISSUE TITLE
Tensorflow takes all my CPUs

### ISSUE THREAD

**Issue by fabienbaradel** [tensorflow/tensorflow#7521]
I am having issue with CPU usage. I am running my code on a node with 4 GPUs and 12 CPUs.
I am using TFRecords for reading my data and it works well. However when I run my model Tensorflow uses all the CPUs available on my node (_even when I specify that I want to use only one thread_):
<img width="587" alt="capture d ecran 2017-02-15 a 10 26 58" src="https://cloud.githubusercontent.com/assets/11565723/22968352/5b682558-f36a-11e6-9bef-97b1e8e4caff.png">

I wanted to check what are the devices founded by Tensorflow by running this:
    **from tensorflow.python.client import device_lib
    print(device_lib.list_local_devices())**

Log:
[name: "/cpu:0"
device_type: "CPU"
memory_limit: 268435456
locality {
}
incarnation: 5188106746934734819
, name: "/gpu:0"
device_type: "GPU"
memory_limit: 12051264308
locality {
  bus_id: 1
}
incarnation: 5913495995344792631
physical_device_desc: "device: 0, name: GeForce GTX TITAN X, pci bus id: 0000:01:00.0"
]

Given the output it seems Tensorflow finds and uses only one CPU. But maybe this CPU founded is a kind of "concatenation" of the 12 CPUs available on my node..

**How to specify Tensorflow to use only one CPU?
How to make Tensorflow understand that I have several CPUs on my node? ( I can't find an config option or something else in the tf.Session()**

Thanks for your help.

I installed Tensorflow using "pip install tensorflow" and below is the output of python -c "import tensorflow; print(tensorflow.__version__)":
**0.12.0-rc1**

An when I run this, it tells me that I have my 12 cores:
import multiprocessing
multiprocessing.cpu_count()
**12**

**Comment by ivan-aksamentov** [comment]
[How can I run Tensorflow on one single core?](http://stackoverflow.com/questions/38187808/how-can-i-run-tensorflow-on-one-single-core)

**Comment by aselle** [comment]
@fabienbaradel, @ivan-aksamentov's response seems to answer your question. Could you please close the issue if it is resolved. Thanks and thanks to @ivan-aksamentov for the quick response.

**Comment by fabienbaradel** [comment]
Thank @ivan-aksamentov for you quick answer!
However your solution doesn't work doesn't work in my case.. @aselle 
I think Tensorflow understands that there is one Cpu on the node which is the "concatenation" of all the CPUs. That is why the I print print(device_lib.list_local_devices()) Tensorflow shows only one CPU.

**Is there a Tensorflow function for checking who many CPUs are available?** (like import multiprocessing;
multiprocessing.cpu_count()). I think the problem comes from there. 

Or maybe it is just a path variable or some config on my server I have to change?

Offtopic: I use iTerm2 ad ZSH_THEME="avit" 
Feel free to find a very good tutorial [here](https://gist.github.com/kevin-smets/8568070)

**Comment by aselle** [comment]
@mrry, do you have any ideas? You could try using perf top to find out what those threads are actually doing. Perhaps it is something else you are doing rather than tensorflow itself?

**Comment by mrry** [comment]
It's impossible to say for certain without the user program. One possibility is that you have many Python threads that could be doing a non-trivial amount of work (e.g. copying buffers for feeding/fetching) outside the GIL and hence in parallel. These threads would be created outside the `tf.Session` and so would not be limited by the configuration options. Some operations (e.g. `ReaderRead`) create their own threads internally as well, and these are not bound by the configuration options either.

TL;DR: If you want to force TensorFlow to use a single core, the best plan right now would be to use OS-level enforcement mechanisms (e.g. cgroups) to limit the process to (and schedule all its threads on) a single core.

**Comment by fabienbaradel** [comment]
Ok thanks for your explanations. I'll try to fix it using your advices. I close the issue and would re-open it if it still doesn't work.

**Comment by noahbishop** [comment]
@fabienbaradel Do you mind sharing how you fixed the issue?

**Comment by fabienbaradel** [comment]
@noahbishop  Have a look at that: https://stackoverflow.com/questions/38187808/how-can-i-run-tensorflow-on-one-single-core

**Comment by byronyi** [comment]
Or use nvidia-docker to run TF and pass the ``--cpus`` option when launching your containers.

**Comment by jayantsharma** [comment]
@byronyi's solution works as far as total cpu time is concerned. I found the `--cpuset-cpus` option more appropriate for my use case to limit the cores docker/tf uses.

**Comment by farnazJz** [comment]
I wanted to occupy only 5 cores out of 40, and to do this I set the  session configuration as follows:

`session_conf = tf.ConfigProto( intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)`

`sess = tf.Session(config=session_conf)`
 
 Try it and if you had any problem, let me know.

**Comment by farnazJz** [comment]
@ivan-aksamentov  
When you run `htop` command on your terminal, your terminal will display those colors. :)

**Comment by ribal-aladeeb** [comment]
> `session_conf = tf.ConfigProto( intra_op_parallelism_threads=1, inter_op_parallelism_threads=1)`

I get
AttributeError: module 'tensorflow' has no attribute 'ConfigProto'
